### PR TITLE
feat(ssh): promoting native extension

### LINKED ISSUE #798: SSH support

### PR DESCRIPTION
**Pull request by squakez**
Closes #798

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    @BuildStep
+    void secureRandomConfiguration(BuildProducer<RuntimeReinitializedClassBuildItem> reinitialized) {
+        for (String s : Arrays.asList(
+                "java.security.SecureRandom",
+                "org.bouncycastle.crypto.CryptoServicesRegistrar",
+                "org.bouncycastle.jcajce.provider.drbg.DRBG$NonceAndIV",
+                "org.bouncycastle.jcajce.provider.drbg.DRBG$Default")) {
+            reinitialized.produce(new RuntimeReinitializedClassBuildItem(s));
+        }
+    }
```
Comment (ppalaga) [inline extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java:59]: I think `org.apache.camel.quarkus.support.bouncycastle.deployment.BouncycastleSupportProcessor` would be the right home for this BuildStep.

File: extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        return new RuntimeInitializedClassBuildItem(EdDSASecurityProviderUtils.class.getCanonicalName());
+    }
+
+    @BuildStep
+    void registerForReflection(BuildProducer<ReflectiveClassBuildItem> reflectiveClasses) {
+        reflectiveClasses.produce(
+                new ReflectiveClassBuildItem(
+                        true,
+                        false,
+                        KeyPairGenerator.class.getCanonicalName(),
```
Comment (ppalaga) [inline extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java:72]: Please prefer `class.getName()` here and also in other methods of this class. It works also for inner classes and is a safer choice for people copying from here.

File: extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import java.security.PublicKey;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import org.apache.sshd.common.util.buffer.Buffer;
+import org.apache.sshd.common.util.security.SecurityUtils;
+
+/**
+ * We're substituting those offending methods that would require the presence of
+ * net.i2p.crypto:eddsa library which is not supported by Camel SSH component
```
Comment (ppalaga) [inline extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java:30]: Is this documented somewhere? I am not finding any mention of eddsa on https://camel.apache.org/components/latest/ssh-component.html Shouldn't we file a documentation issue?

File: integration-tests/ssh/src/test/java/org/apache/camel/quarkus/component/ssh/it/SshTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@QuarkusTest
+@QuarkusTestResource(SshTestResource.class)
+class SshTest {
+
+    @Test
+    public void testWriteToSSHAndReadFromSSH() {
+        final String fileContent = "Hello Camel Quarkus SSH";
+        BouncyCastleProvider x;
```
Comment (ppalaga) [inline integration-tests/ssh/src/test/java/org/apache/camel/quarkus/component/ssh/it/SshTest.java:35]: Can't we remove this line?

**Review by ppalaga — COMMENTED** [review]
Nice work, thanks @squakez! Some comments and suggestions inline.

File: extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java (hunk 1 above)
Comment (squakez) [reply extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java:59]: Yeah, I was in doubt if it belongs there. Thanks for confirming, I am going to move it accordingly.

File: extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java (hunk 2 above)
Comment (lburgazzoli) [reply extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java:72]: you may also remove the `.getName()` AFAIK

File: extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java (hunk 3 above)
Comment (squakez) [reply extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java:30]: I think I did not explain properly in the comment. The `camel-ssh` depends on `sshd-core` library which has 2 optional security providers, `bouncycastle` and `eddsa`. In `camel-ssh` we chose to use `bouncycastle`. I will rephrase the comment to highlight that.

File: extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java (hunk 2 above)
Comment (ppalaga) [reply extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java:72]: True, ReflectiveClassBuildItem accepts also `Class...` for which it calls `getName()` under the hood.

Comment (squakez) [reply extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java:72]: > you may also remove the `.getName()` AFAIK

Do you mean using a text with the fully qualified name instead?

File: extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java (hunk 3 above)
Comment (ppalaga) [reply extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java:30]: > I think I did not explain properly in the comment. The `camel-ssh` depends on `sshd-core` library which has 2 optional security providers, `bouncycastle` and `eddsa`. In `camel-ssh` we chose to use `bouncycastle`.

Thanks for the clarification. I still think that this is a piece of info interesting for the end users and it should ideally be present on https://camel.apache.org/components/latest/ssh-component.html . Filing a documentation issue would be a way not forget about that :)

File: extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java (hunk 2 above)
Comment (squakez) [reply extensions/ssh/deployment/src/main/java/org/apache/camel/quarkus/component/ssh/deployment/SshProcessor.java:72]: Ah, okey, I see, thanks!

File: extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java (hunk 3 above)
Comment (squakez) [reply extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java:30]: Okey, I'll take care!

Comment (squakez) [reply extensions/ssh/runtime/src/main/java/org/apache/camel/quarkus/component/ssh/runtime/SubstituteSecurityUtils.java:30]: For the record: https://issues.apache.org/jira/browse/CAMEL-15550

**Review by lburgazzoli — APPROVED** [review]

**Review by ppalaga — APPROVED** [review]

**Comment by squakez** [comment]
The failing check does not seem related to this PR, any possibility to rerun it?

**Comment by jamesnetherton** [comment]
I think it's fine to ignore the failing test. Lets get this merged.

**Pull request merged**